linera-execution/src/system.rs

```diff
@@ -976,3 +983,180 @@ where
+        let context = OperationContext {
+            chain_id,
+            authenticated_signer: None,
+            height,
+            index: operation_index,
+            next_message_index,
+        };
+        let operation = SystemOperation::PublishBytecode {
+            contract: Bytecode::new(vec![]),
+            service: Bytecode::new(vec![]),
+        };
+        let (result, new_application) = view
+            .system
+            .execute_operation(&context, &operation)
+            .await
+            .unwrap();
+        assert_eq!(new_application, None);
+        assert_eq!(
+            result.messages[PUBLISH_BYTECODE_MESSAGE_INDEX as usize].message,
+            SystemMessage::BytecodePublished { operation_index }
+        );
+    }
+
+    #[tokio::test]
+    async fn application_message_index() {
+        let description = ChainDescription::Root(5);
+        let chain_id = ChainId::from(description);
+        let state = SystemExecutionState {
+            description: Some(description),
+            ..SystemExecutionState::default()
+        };
+        let mut view =
+            ExecutionStateView::<MemoryContext<TestExecutionRuntimeContext>>::from_system_state(
+                state,
+            )
+            .await;
+
+        let height = BlockHeight::from(7);
+        let next_message_index = 8;
+        let operation_index = 2;
+        let context = OperationContext {
+            chain_id,
+            authenticated_signer: None,
+            height,
+            index: operation_index,
+            next_message_index,
+        };
+
+        let bytecode_id = BytecodeId::new(MessageId {
+            chain_id,
+            height: BlockHeight::from(5),
+            index: 0,
+        });
+        let location = BytecodeLocation {
+            certificate_hash: CryptoHash::new(&Dummy),
+            operation_index: 1,
+        };
+        view.system
+            .registry
+            .register_published_bytecode(bytecode_id, location)
+            .unwrap();
+
+        let operation = SystemOperation::CreateApplication {
+            bytecode_id,
+            parameters: vec![],
+            initialization_argument: vec![],
+            required_application_ids: vec![],
+        };
+        let (result, new_application) = view
+            .system
+            .execute_operation(&context, &operation)
+            .await
+            .unwrap();
+        assert_eq!(
+            result.messages[CREATE_APPLICATION_MESSAGE_INDEX as usize].message,
+            SystemMessage::ApplicationCreated
+        );
+        let creation = MessageId {
+            chain_id,
+            height,
+            index: next_message_index + CREATE_APPLICATION_MESSAGE_INDEX,
+        };
+        let id = ApplicationId {
+            bytecode_id,
+            creation,
+        };
+        assert_eq!(new_application, Some((id, vec![])));
+    }
+
+    #[tokio::test]
+    async fn open_chain_message_index() {
```

Comment:
Can we share the boilerplate between the N similar tests? (since N will keep growing)